[GVNSink] Fix non-determinisms by using a deterministic ordering

hiraditya opened the pull request:
GVNSink used to order instructions based on their pointer values and was prone to non-determinism because of that. 
This patch ensures all the values stored are using a deterministic order. I have also added a verfier(`ModelledPHI::verifyModelledPHI`) to assert when ordering isn't preserved.

Additionally, I have added a test case that is a mirror image of an existing test.

Fixes: #77852

hiraditya committed:
Fix non-determinisms

GVNSink used to order instructions based on their pointer values and was prone to
non-determinism because of that. Using DFSNumber for ordering all the values
fixes the non-determinism

hiraditya commented on llvm/test/Transforms/GVNSink/int_sideeffect.ll:
```diff
@@ -28,3 +28,27 @@ if.end:
   ret float %phi
 }
 
+; CHECK-LABEL: scalarsSinkingReverse
```
This test is just a reverse of the test above to ensure GVNSink doesn't depend on the order of branches.

ilovepi replied:
can you add that detail as a comment in the test. This seems like something that could get lost and cause the test to be removed.

ilovepi reviewed (commented):
This seems like it should fix the issue to me, but lets see if @nikic agrees.

efriedma-quic commented on llvm/lib/Transforms/Scalar/GVNSink.cpp:
```diff
@@ -222,20 +222,29 @@ raw_ostream &operator<<(raw_ostream &OS, const SinkingInstructionCandidate &C) {
-    llvm::sort(Ops);
+
+    auto DFSOrder = [DT](OpsType O1, OpsType O2) {
+      return DT->getNode(O1.first) < DT->getNode(O2.first);
```
DT->getNode() returns a pointer.  If you want the domtree to generate DFS numbers, you need to explicitly request them (see updateDFSNumbers()/getDFSNumIn()).

Splitting an edge with SplitBlockPredecessors will invalidate this numbering; not sure how you handle that.

hiraditya replied:
Ah, that was the intent. Thanks for pointing it out.

efriedma-quic commented on llvm/lib/Transforms/Scalar/GVNSink.cpp:
```diff
@@ -886,8 +919,12 @@ void GVNSink::sinkLastInstruction(ArrayRef<BasicBlock *> Blocks,
   if (!G.run(F))
     return PreservedAnalyses::all();
+
+  // PHI nodes get inserted which haven't been added to the Dominator Tree.
```
PHI nodes aren't represented in the dominator tree.  The issue is the call to SplitBlockPredecessors().  (You probably could preserve the domtree without much effort; SplitBlockPredecessors() has code to preserve it if you pass in a DomTreeUpdater.)

hiraditya replied:
That makes sense. I see the code around `SplitBlockPredecessors`  has no test. And i'm not sure if this is a useful case either. I'm fine removing that part of code altogether. WDYT?

efriedma-quic replied:
That seems like something you could measure pretty easily.

hiraditya replied:
nvm. I found a way to update it. Posting the patch soon.

hiraditya committed:
Fix issues with DT update, use DFSNumbers

efriedma-quic reviewed (commented):
SplitBlockPredecessors invalidates the DFS numbering, I think.

Also, SplitBlockPredecessors changes the predecessor list of a basic block, which I think needs to invalidate any ModelledPHI referring to that block?

efriedma-quic commented:
> > SplitBlockPredecessors invalidates the DFS numbering, I think.
> 
> Makes sense to just update DFSNumbes with each split block.

updateDFSNumbers walks the whole function, so that's quadratic in the size of the function.  If you don't care about the exact numbers, maybe consider just building your own numbering as you visit PHI nodes.  It's only a few lines to construct a `DenseMap<BasicBlock*, int>`.

> > Also, SplitBlockPredecessors changes the predecessor list of a basic block, which I think needs to invalidate any ModelledPHI referring to that block?
> 
> AFAICT all the ModelledPHI are created for a specific SinkBB (in SinkBB and analyzeInstructionForSinking functions) and then destroyed.

So the numbers don't need to be stable across blocks?  That makes things simpler.

hiraditya committed:
Populate DFS numbers ahead of time

This prevents quadratic behavior of calling DominatorTree::updateDFSNumbers everytime
BasicBlocks are split to sink instructions

efriedma-quic commented on llvm/lib/Transforms/Scalar/GVNSink.cpp:
```diff
@@ -575,6 +607,13 @@ class GVNSink {
+    // when CFG is modified. The DFSNumbers of newly created basic blocks
+    // are irrelevant because RPOT is also obtained ahead of time and only
+    // DFSNumbers of original CFG are relevant for sinkable candidates.
+    for (auto *BB : RPOT)
```
At this point, do you actually need DFS numbering?  If any arbitrary numbering works, might as well just number in IR order.

hiraditya replied:
That's true, it'll simplify a lot of things as you noted. I'll update the patch.

efriedma-quic commented on llvm/lib/Transforms/Scalar/GVNSink.cpp:
```diff
@@ -766,6 +807,9 @@ unsigned GVNSink::sinkBB(BasicBlock *BBEnd) {
              BBEnd->printAsOperand(dbgs()); dbgs() << "\n");
   SmallVector<BasicBlock *, 4> Preds;
   for (auto *B : predecessors(BBEnd)) {
+    // Bailout on malformed CFG where BasicBlock has no predecessor(PR42346).
```
Malformed is an exaggeration; the IR verifier is perfectly fine with code like this.

Realistically, it'll almost always get cleaned up before this pass runs, so it doesn't really matter what you do here, but please use clear wording.

hiraditya replied:
will do.

nikic commented on llvm/lib/Transforms/Scalar/GVNSink.cpp:
```diff
@@ -575,6 +607,13 @@ class GVNSink {
+    // DFSNumbers of original CFG are relevant for sinkable candidates.
+    for (auto *BB : RPOT)
+      if (DT->getNode(BB))
+        DFSNumbers[BB] = DT->getNode(BB)->getDFSNumIn();
```
Do you actually care about these being DFS numbers, rather than some arbitrary block numbering? If not, you could just use the numbering produced by RPOT.

hiraditya replied:
True, Eli has also noted in a previous comment. I'll update the patch.

hiraditya committed:
Use reverse-post order instead of DFS order

As noted by Eli and Nikita, there is no need to use depth first ordering
because all we need is a deterministic order. This approach avoids calling expensive
DFSNumbering operation and reuse an existing visitation order

efriedma-quic commented on llvm/lib/Transforms/Scalar/GVNSink.cpp:
```diff
@@ -607,13 +606,16 @@ class GVNSink {
-    // are irrelevant because RPOT is also obtained ahead of time and only
-    // DFSNumbers of original CFG are relevant for sinkable candidates.
+    // Populate reverse post-order to order basic blocks in deterministic
+    // order. Any arbitray ordering will work in this case as long they are
```
```suggestion
    // order. Any arbitrary ordering will work in this case as long they are
```

efriedma-quic reviewed (approved):
LGTM
